### ISSUE TITLE
Require PHPUnit 4.8 in composer.json

### ISSUE THREAD

**Issue by SamMousa** [twigphp/Twig#2085]
The required version of PHPUnit should be listed as dev dependency in `composer.json`.

The tests will give a lot of deprecation notices when run with PHPUnit 5.

**Comment by stof** [comment]
they give deprecations only on PHPUnit 5.4, not on PHPUnit 5.3, and this is because the testsuite has not been adapted yet for the PHPunit 5.4 deprecations.

Requiring PHPUnit 4.8 would be wrong. PHPunit 4.8 is not officially supported on PHP 7+

**Comment by SamMousa** [comment]
Ah, didnt think 5.3 to 5.4 made the difference; regardless of what version,
shouldn't it be required in composer? To avoid exactly this kind of issue :p

**Comment by SamMousa** [comment]
@stof, why do you think @fabpot  would not want to use `require-dev`?
Isn't that exactly what it is for? Requiring development dependencies?

**Comment by stof** [comment]
Well, he already rejected this in the past, saying PHPUnit is best installed globally with the phar.

and btw, adding it as a dev requirement will force to install approx 15 additional packages on Travis for each build, while it already has PHPUnit installed.

a much better contribution would be to update the testsuite to run without calling deprecated APIs (but keeping compat with PHPUnit 4.8 as we need to run tests on PHP 5.5)

**Comment by SamMousa** [comment]
@stof, whether he rejected it in the past or not is not really relevant; although it could be an indicator ;-)
However when we are requiring a version of PHPUnit that is not the latest, the global argument doesn't make a lot of sense...
I contribute to a lot of projects and if several of them tell me to install my dependencies globally I run into problems, which is exactly why global dependencies are considered bad.

https://www.sitepoint.com/composer-global-require-considered-harmful/

It references a discussion where you replied to as well btw ;-)

**Comment by SamMousa** [comment]
It seems this is a hot topic though: https://github.com/symfony/symfony/issues/12136 and the travis argument comes up quite a lot.

**Comment by stof** [comment]
@SamMousa I never said it should be installed with ``composer global require``. The proper way to install a tool globally is to install it separately from other global tools. And this is very easy for PHPUnit as they distribute it as a Phar.

And as I said, the testsuite has not been updated to avoid PHPUnit 5.4 deprecations **yet** (because nobody worked on it yet). But this does not even make it impossible to run the testsuite with 5.4 (it just gives you a bunch of additional warnings).

**Comment by SamMousa** [comment]
Yes I know, those are valid arguments; the problem I have is that I need to figure out wat version of these tools to use.
Twig supports (according to composer.json) PHP 5.2..
However PHP 5.2, 5.3 and 5.4 are not supported by PHPUnit 5.0.

https://github.com/sebastianbergmann/phpunit/wiki/Release-Announcement-for-PHPUnit-5.0.0

So technically there is not even a PHPUnit version we can use to test PHP 5.2 and PHP 7..

**Comment by SamMousa** [comment]
Travis CI runs PHPUnit 3.6 -5.4, depending on the PHP version it is testing; at some point it will not be possible to have 1 set of tests that supports all PHPUnit versions...

**Comment by SamMousa** [comment]
I looked into what needs to be done to remove the deprecation notices..
````
PHPUnit_Framework_TestCase::getMock() is deprecated, use PHPUnit_Framework_TestCase::createMock() or PHPUnit_Framework_TestCase::getMockBuilder()
````

However, `createMock` is available since 5.4.0. `getMockBuilder` is available since 3.5.0, but refers to a class in a dependency `phpunit/phpunit-mock-objects`, where there have been 2 major releases between the version used by 3.5.0 and 5.4.0...

In short, there is no way to guarantee it is possible to create a single PHPUnit test that works with all past and future versions of PHPUnit...

**Comment by stof** [comment]
@SamMousa ``getMockBuilder`` is available in all the PHPUnit versions we care about (it has not been replaced when replacing ``getMock``).
and btw, a bunch of tests are already using ``getMockBuilder`` when they need more control over the created mock.

**Comment by SamMousa** [comment]
@stof I know, but the assumption here is that the `PHPUnit_Framework_MockObject_MockBuilder` class has not / will not change, even though the dependency version used ranges from `1.2` to `3.1`...
Even if that assumption holds for now (which I believe it does), it is not something that you can depend on.

Essentially at some point the test would need to be tested to make sure they work with all PHPUnit versions even ones that have not been supported for several years; anyway, this is drifting a bit off topic.

In conclusion what you're saying is, the tests should with the latest version of PHPUnit and developers / contributors should thus use the latest PHPUnit version when running the tests?

**Comment by SamMousa** [comment]
> a much better contribution would be to update the testsuite to run without calling deprecated APIs (but keeping compat with PHPUnit 4.8 as we need to run tests on PHP 5.5)

https://github.com/twigphp/Twig/pull/2087
